File: impl/tools/dsipvec/envelope.py
"""DSIP-JOSE envelope: construct, sign, and the verification pipeline (spec §10.2, §7.4, §12.9, §20.6).

Pipeline stages 1–11 of vectors/README.md. The payload is carried as bytes
through signature verification and only then decoded — never re-serialized.
"""
from __future__ import annotations

import json
from dataclasses import dataclass, field
from typing import Any

from .crypto import (KeyPair, b64url_decode, b64url_encode, ed25519_verify,
                     public_from_did_key, b58_decode, ED25519_PUB_MULTICODEC)
from .verdict import Verdict
from .wire import parse_payload, DID_RE
from . import ulid as ulid_mod

REPLAY_WINDOW_S = 300          # §12.9
ULID_TOLERANCE_S = 300         # §20.6, Impl: tolerance = replay window (spec-gap 6)
INTRODUCTION_MAX_VALIDITY_S = 604800  # §19.4; Impl: enforced at stage 9 (spec-gap 31)
WS_MAX_ENVELOPE_BYTES = 65536  # §13.2
SIGNALING_CAPABILITY = "dsip.signaling"


# ---------------------------------------------------------------- construction

def protected_header(kid: str, alg: str = "EdDSA", **extra) -> dict:
    h = {"alg": alg, "kid": kid, "typ": "dsip+json"}
    h.update(extra)
    return h


def encode_payload(payload: dict) -> bytes:
    """Compact UTF-8 JSON. Field order is whatever the dict holds; the signature covers these bytes."""
    return json.dumps(payload, separators=(",", ":"), ensure_ascii=False).encode("utf-8")


def sign_bytes(payload_bytes: bytes, signer: KeyPair, kid: str, header_extra: dict | None = None,
               alg: str = "EdDSA") -> dict:
    header = protected_header(kid, alg, **(header_extra or {}))
    prot = b64url_encode(encode_payload(header))
    pay = b64url_encode(payload_bytes)
    sig = signer.sign(f"{prot}.{pay}".encode("ascii"))
    return {"protected": prot, "payload": pay, "signature": b64url_encode(sig)}


def sign(payload: dict, signer: KeyPair, kid: str, header_extra: dict | None = None, alg: str = "EdDSA") -> dict:
    return sign_bytes(encode_payload(payload), signer, kid, header_extra, alg)


def frame(envelope: dict) -> str:
    """The ws/1.0 text frame for an envelope (§13.2 framing)."""
    return json.dumps(envelope, separators=(",", ":"))


# ---------------------------------------------------------------- resolution (§8.1)

@dataclass
class Context:
    now: int
    did_documents: dict[str, dict] = field(default_factory=dict)
    delegations: list[dict] = field(default_factory=list)
    seen_ids: set[str] = field(default_factory=set)
    supported: dict[str, Any] = field(default_factory=lambda: {"core": "1.0", "profiles": [], "extensions": []})
    revocations: list = field(default_factory=list)

    @staticmethod
    def from_vector(ctx: dict) -> "Context":
        return Context(
            now=ctx.get("now", 0),
            did_documents=ctx.get("did_documents", {}) or {},
            delegations=list(ctx.get("delegations", []) or []),
            seen_ids=set(ctx.get("seen_ids", []) or []),
            supported=ctx.get("supported") or {"core": "1.0", "profiles": ["interactive-media/1.0"], "extensions": []},
            revocations=list(ctx.get("revocations", []) or []),
        )


def split_kid(kid: str) -> tuple[str, str] | None:
    if not isinstance(kid, str) or "#" not in kid:
        return None
    did, frag = kid.split("#", 1)
    if not DID_RE.match(did) or frag == "":
        return None
    return did, frag


def _multibase_to_ed25519(mb: str) -> bytes | None:
    if not isinstance(mb, str) or not mb.startswith("z"):
        return None
    try:
        raw = b58_decode(mb[1:])
    except ValueError:
        return None
    if raw[:2] != ED25519_PUB_MULTICODEC or len(raw) != 34:
        return None
    return raw[2:]


def resolve_kid(kid: str, ctx: Context) -> bytes | None:
    """DID URL → Ed25519 public key, or None. Authority: DID document only (§8.1 rule 4)."""
    parts = split_kid(kid)
    if parts is None:
        return None
    did, frag = parts
    if did.startswith("did:key:"):
        # Self-certifying: the only verification method is the key itself; fragment must name it.
        if frag != did[len("did:key:"):]:
            return None
        try:
            return public_from_did_key(did)
        except ValueError:
            return None
    doc = ctx.did_documents.get(did)
    if not isinstance(doc, dict):
        return None
    for vm in doc.get("verificationMethod", []) or []:
        if not isinstance(vm, dict):
            continue
        vid = vm.get("id")
        if vid == kid or vid == f"#{frag}":
            if vm.get("controller") not in (None, did):
                return None
            return _multibase_to_ed25519(vm.get("publicKeyMultibase"))
    return None


# ---------------------------------------------------------------- raw verify (stages 2–6)

@dataclass
class Verified:
    header: dict
    payload: dict
    payload_bytes: bytes
    signer_did: str
    kid: str


def verify_raw(envelope: Any, ctx: Context, core_shape: bool = True) -> tuple[Verdict, Verified | None]:
    """Stages 2–6: shape, header, kid resolution, signature, payload parse. No binding, no timing.

    `core_shape=False` skips the DSIP core-field check (used for delegation credentials,
    whose payload is a DeviceDelegation object, not a message)."""
    if not isinstance(envelope, dict) or set(envelope.keys()) != {"protected", "payload", "signature"}:
        return Verdict.reject("envelope-shape"), None
    try:
        prot_b, pay_b, sig_b = (b64url_decode(envelope[k]) if isinstance(envelope[k], str) else None
                                for k in ("protected", "payload", "signature"))
    except ValueError:
        return Verdict.reject("envelope-shape"), None
    if prot_b is None or pay_b is None or sig_b is None:
        return Verdict.reject("envelope-shape"), None
    try:
        header = json.loads(prot_b.decode("utf-8"))
    except (ValueError, UnicodeDecodeError):
        return Verdict.reject("header-invalid"), None
    if not isinstance(header, dict) or not isinstance(header.get("alg"), str) or not isinstance(header.get("kid"), str):
        return Verdict.reject("header-invalid"), None
    if header["alg"] != "EdDSA":
        # §10.2: Ed25519 MUST; ES256 MAY (not implemented here); everything else MUST be rejected.
        return Verdict.reject("alg-unsupported"), None
    parts = split_kid(header["kid"])
    if parts is None:
        return Verdict.reject("kid-invalid"), None
    pub = resolve_kid(header["kid"], ctx)
    if pub is None:
        return Verdict.reject("kid-unresolvable"), None
    signing_input = f"{envelope['protected']}.{envelope['payload']}".encode("ascii")
    if len(sig_b) != 64 or not ed25519_verify(pub, signing_input, sig_b):
        return Verdict.reject("signature-invalid"), None
    v, payload = parse_payload(pay_b, core_shape=core_shape)
    if not v.ok:
        return v, None
    return Verdict.accept(), Verified(header, payload, pay_b, parts[0], header["kid"])


# ---------------------------------------------------------------- delegation (§7.4)

def verify_delegation(deleg: Any, subject: str, device: str, ctx: Context,
                      capability: str = SIGNALING_CAPABILITY) -> Verdict:
    """A delegation envelope is valid for (subject, device) when it is signed directly by a key
    of `subject`, names exactly that subject/device, carries `capability` (dsip.signaling for
    envelopes; dsip.messaging for MLS leaves, Messaging Profile M§6.2), and is live at now."""
    v, ver = verify_raw(deleg, ctx, core_shape=False)
    if not v.ok:
        return Verdict.reject("delegation-invalid", detail=v.code)
    p = ver.payload
    if (p.get("type") != "DeviceDelegation" or p.get("subject") != subject or p.get("device") != device
            or ver.signer_did != subject):
        return Verdict.reject("delegation-invalid")
    caps = p.get("capabilities")
    if not isinstance(caps, list) or capability not in caps:
        return Verdict.reject("delegation-capability")
    ia, ea = p.get("issued_at"), p.get("expires_at")
    if not (isinstance(ia, int) and isinstance(ea, int)) or not (ia <= ctx.now < ea):
        return Verdict.reject("delegation-expired")
    if any(revokes(r, subject, device, ia, ctx) for r in revocations_for(subject, ctx)):
        return Verdict.reject("delegation-revoked")  # spec-gap 57 (v0.8 draft)
    return Verdict.accept()


REVOCATIONS_DOC_FIELD = "dsipDelegationRevocations"


def _as_envelope(r):
    if isinstance(r, str):
        parts = r.split(".")
        return {"protected": parts[0], "payload": parts[1], "signature": parts[2]} if len(parts) == 3 else None
    return r if isinstance(r, dict) else None


def revocations_for(subject: str, ctx: Context) -> list:
    """Every revocation a verifier can see for `subject`: those it holds, and those the subject's DID document
    publishes (authoritative, §8.1). A validly signed revocation only removes authority, so any source counts."""
    doc = ctx.did_documents.get(subject)
    published = doc.get(REVOCATIONS_DOC_FIELD, []) if isinstance(doc, dict) else []
    return [e for e in map(_as_envelope, list(ctx.revocations) + list(published or [])) if e is not None]


def revokes(rev: dict, subject: str, device: str, delegation_issued_at: int, ctx: Context) -> bool:
    """spec-gap 57: a `delegation-revocation` signed directly by a key of `subject` revokes every delegation of
    `device` for `subject` issued at or before its `revoked_at`. The record is a credential: no delivery window."""
    v, ver = verify_raw(rev, ctx, core_shape=False)
    if not v.ok:
        return False
    p = ver.payload
    return (p.get("type") == "delegation-revocation" and p.get("subject") == subject and p.get("from") == subject
            and ver.signer_did == subject and p.get("device") == device and isinstance(p.get("revoked_at"), int)
            and delegation_issued_at <= p["revoked_at"])


def check_binding(subject: str, device: str, presented: list, ctx: Context) -> Verdict:
    """Is `device` authorized to act for `subject`? Direct when equal; else via a presented delegation."""
    if subject == device:
        return Verdict.accept()
    candidates = [d for d in presented if isinstance(d, dict) and _names(d) == (subject, device)]
    if not candidates:
        return Verdict.reject("signer-mismatch")
    first_failure = None
    for d in candidates:
        v = verify_delegation(d, subject, device, ctx)
        if v.ok:
            return v
        first_failure = first_failure or v
    return first_failure


def _names(deleg: dict) -> tuple | None:
    try:
        p = json.loads(b64url_decode(deleg["payload"]).decode("utf-8"))
        return (p.get("subject"), p.get("device"))
    except Exception:
        return None


# ---------------------------------------------------------------- full pipeline (stages 1–11)

def verify(envelope: Any, ctx: Context, frame_text: str | None = None) -> tuple[Verdict, Verified | None]:
    """Stages 1–11. On accept, extra carries type/signer/identity."""
    if frame_text is not None and len(frame_text.encode("utf-8")) > WS_MAX_ENVELOPE_BYTES:
        return Verdict.reject("frame-too-large", "transport.envelope-too-large"), None
    v, ver = verify_raw(envelope, ctx)
    if not v.ok:
        return v, None
    p = ver.payload
    msg_type = p["type"]
    hello_reason = "transport.hello-rejected" if msg_type == "hello" else None

    # Stage 7: bind kid → from (→ on_behalf_of on hello)
    presented = list(ctx.delegations) + [d for d in (ver.header.get("delegations") or []) if isinstance(d, dict)]
    b = check_binding(p["from"], ver.signer_did, presented, ctx)
    if not b.ok:
        return Verdict.reject(b.code, hello_reason, b.detail), None
    identity = p["from"]
    if msg_type == "hello" and isinstance(p.get("on_behalf_of"), str):
        b = check_binding(p["on_behalf_of"], p["from"], presented, ctx)
        if not b.ok:
            return Verdict.reject(b.code, hello_reason, b.detail), None
        identity = p["on_behalf_of"]

    # Stage 8–9: expiry ordering, replay window, expiry
    ia, ea = p["issued_at"], p["expires_at"]
    if ea <= ia:
        return Verdict.reject("expiry-order"), None
    if msg_type == "introduction":
        # §19.4: introductions are store-and-forward by design, valid up to 604,800 s. Impl (spec-gap 31):
        # the cap is enforced here, the 300 s age bound is replaced by expires_at (checked next), the
        # future bound stays, and receivers track the id until expires_at.
        if ea - ia > INTRODUCTION_MAX_VALIDITY_S:
            return Verdict.reject("introduction-validity"), None
        if ia > ctx.now + REPLAY_WINDOW_S:
            return Verdict.reject("replay-window"), None
    elif ia < ctx.now - REPLAY_WINDOW_S or ia > ctx.now + REPLAY_WINDOW_S:
        return Verdict.reject("replay-window"), None
    if ea < ctx.now:
        return Verdict.reject("expired", "session.expired" if msg_type == "invite" else None), None
    # Stage 10: dedup
    if p["id"] in ctx.seen_ids:
        return Verdict.reject("duplicate-id"), None
    # Stage 11: ULID timestamp vs issued_at (§20.6)
    if abs(ulid_mod.timestamp_ms(p["id"]) // 1000 - ia) > ULID_TOLERANCE_S:
        return Verdict.reject("ulid-issued-at-mismatch"), None
    return Verdict.accept(type=msg_type, signer=ver.signer_did, identity=identity), ver

File: impl-ts/src/envelope.ts
/**
 * The DSIP-JOSE envelope verification pipeline: from received bytes to a verified payload.
 *
 * Spec: §10.2 (signature over the transmitted bytes), §10.3 (payload rules), §7.4 (delegation and
 * revocation), §12.9 (replay window), §20.6 (ULID/`issued_at`), §13.2 (frame cap, hello first).
 * The stage order is the conformance suite's (`impl/vectors/README.md`, "Pipeline order").
 */
import { createPublicKey, verify as cryptoVerify } from "node:crypto";
import { didOf, isDidUrlWithFragment, resolveKey, type DidDocuments, type Json, type JsonObject } from "./did.js";
import { b64urlDecode, hasFloat, isUlid, ulidSeconds, utf8Decode } from "./encoding.js";
import type { SchemaSet } from "./schema.js";
import { checkPayload, type SemanticContext } from "./semantic.js";
import { reject, type Reject, type Verdict } from "./verdict.js";

/** Spec: §13.2 — a ws/1.0 text frame is at most 65,536 bytes. */
export const MAX_FRAME_BYTES = 65536;
/** Spec: §12.9 — replay window, seconds either side of receipt. */
export const REPLAY_WINDOW_S = 300;
/** Spec: §20.6 — Impl: tolerance between the ULID timestamp and `issued_at` (spec-gap 6). */
export const ULID_TOLERANCE_S = 300;
/** Spec: §19.4 — longest introduction validity, seconds (7 days). */
export const MAX_INTRODUCTION_VALIDITY_S = 604800;

/** A DSIP-JOSE envelope as transmitted. Spec: §10.2 */
export interface Envelope {
  protected: string;
  payload: string;
  signature: string;
}

/** Everything a receiver knows when an envelope arrives. */
export interface ReceiverContext extends SemanticContext {
  /** Receiver clock at receipt, integer seconds. */
  now: number;
  did_documents?: DidDocuments;
  /** Delegations held by the verifier (envelopes). Spec: §7.4 */
  delegations?: Json[];
  /** Delegation revocations held by the verifier (envelopes). Spec: §7.4 (v0.8) */
  revocations?: Json[];
  /** Message ids already seen inside the replay window. Spec: §12.9 */
  seen_ids?: string[];
  /** False while the connection has no verified `hello`. Spec: §13.2 */
  hello_verified?: boolean;
}

const SPKI_ED25519 = Buffer.from("302a300506032b6570032100", "hex");

/** Ed25519 verification over exact bytes. Spec: §10.2 */
function ed25519Verify(key: Buffer, message: Buffer, signature: Buffer): boolean {
  if (signature.length !== 64) return false;
  try {
    const pub = createPublicKey({ key: Buffer.concat([SPKI_ED25519, key]), format: "der", type: "spki" });
    return cryptoVerify(null, message, pub, signature);
  } catch {
    return false;
  }
}

/** An envelope whose signature verified. */
interface Opened {
  header: JsonObject;
  kid: string;
  /** The payload exactly as signed. */
  bytes: Buffer;
}

/** Accept a JSON envelope object or the compact `protected.payload.signature` form. */
function asEnvelope(value: Json | undefined): Envelope | null {
  if (typeof value === "string") {
    const parts = value.split(".");
    return parts.length === 3 ? { protected: parts[0]!, payload: parts[1]!, signature: parts[2]! } : null;
  }
  if (!value || typeof value !== "object" || Array.isArray(value)) return null;
  const keys = Object.keys(value).sort().join(",");
  if (keys !== "payload,protected,signature") return null;
  if (!["protected", "payload", "signature"].every((k) => typeof value[k] === "string")) return null;
  return value as unknown as Envelope;
}

/** Stages 2–5. The payload stays raw bytes until the signature has verified (§10.2). */
function open(value: Json | undefined, documents: DidDocuments): Opened | Reject {
  const env = asEnvelope(value);
  const parts = env && [env.protected, env.payload, env.signature].map(b64urlDecode);
  if (!env || !parts || parts.some((p) => p === null)) return reject("envelope-shape");
  const [headerBytes, bytes, signature] = parts as [Buffer, Buffer, Buffer];

  let header: Json;
  try {
    header = JSON.parse(utf8Decode(headerBytes) ?? "!");
  } catch {
    return reject("header-invalid");
  }
  if (!header || typeof header !== "object" || Array.isArray(header)) return reject("header-invalid");
  if (typeof header["alg"] !== "string" || typeof header["kid"] !== "string") return reject("header-invalid");
  if (header["alg"] !== "EdDSA") return reject("alg-unsupported"); // §10.2: ES256 is MAY; everything else MUST be rejected
  const kid = header["kid"];
  if (!isDidUrlWithFragment(kid)) return reject("kid-invalid");

  const key = resolveKey(kid, documents);
  if (!key) return reject("kid-unresolvable");
  // §10.2: the signing input is the transmitted ASCII text, never a re-serialization.
  const signingInput = Buffer.from(`${env.protected}.${env.payload}`, "ascii");
  if (!ed25519Verify(key, signingInput, signature)) return reject("signature-invalid");
  return { header, kid, bytes };
}

/** Stage 6, first three steps: UTF-8, a JSON object, no floats. Spec: §10.3 */
export function decodePayload(bytes: Uint8Array): JsonObject | Reject {
  const text = utf8Decode(bytes);
  if (text === null) return reject("payload-not-utf8");
  let value: Json;
  try {
    value = JSON.parse(text);
  } catch {
    return reject("payload-not-json");
  }
  if (!value || typeof value !== "object" || Array.isArray(value)) return reject("payload-not-json");
  if (hasFloat(text)) return reject("payload-float");
  return value;
}

function isReject(v: unknown): v is Reject {
  return !!v && typeof v === "object" && (v as Reject).verdict === "reject";
}

/** A verified `delegation-revocation`. */
interface Revocation {
  subject: string;
  device: string;
  revoked_at: number;
}

/**
 * The revocations that count: signed directly by their subject, `from` = `subject`.
 *
 * Spec: §7.4 (v0.8) — verified by signature and subject, not by a delivery window; found in the
 * verifier's store and in the subject's document (`dsipDelegationRevocations`). Invalid ones are ignored.
 */
function revocationsFor(subject: string, ctx: ReceiverContext): Revocation[] {
  const documents = ctx.did_documents ?? {};
  const published = documents[subject]?.["dsipDelegationRevocations"];
  const out: Revocation[] = [];
  for (const candidate of [...(ctx.revocations ?? []), ...(Array.isArray(published) ? published : [])]) {
    const opened = open(candidate, documents);
    if (isReject(opened)) continue;
    const p = decodePayload(opened.bytes);
    if (isReject(p) || p["type"] !== "delegation-revocation") continue;
    if (p["subject"] !== subject || p["from"] !== subject || didOf(opened.kid) !== subject) continue;
    if (typeof p["device"] !== "string" || typeof p["revoked_at"] !== "number") continue;
    out.push({ subject, device: p["device"], revoked_at: p["revoked_at"] });
  }
  return out;
}

/**
 * Bind `device` to `identity` through a delegation; `null` when bound.
 *
 * Spec: §7.4 — a delegation is an envelope signed directly by a key of the subject (no chains),
 * valid when `issued_at ≤ now < expires_at`, carrying `dsip.signaling`; revocation is checked after
 * the signature, capability and validity checks.
 * Impl: delegations the verifier merely holds for other pairs are irrelevant; a delegation
 * *presented* in the protected header that does not link this pair is `delegation-invalid`.
 */
function bind(device: string, identity: string, header: JsonObject, ctx: ReceiverContext): Reject | null {
  const documents = ctx.did_documents ?? {};
  const presented = Array.isArray(header["delegations"]) ? header["delegations"] : [];
  const candidates = [
    ...presented.map((d) => ({ d, presented: true })),
    ...(ctx.delegations ?? []).map((d) => ({ d, presented: false })),
  ];
  let failure: Reject | null = null;
  for (const { d, presented: wasPresented } of candidates) {
    const verdict = checkDelegation(d, device, identity, documents, ctx);
    if (verdict === "bound") return null;
    if (verdict === "unrelated" && !wasPresented) continue;
    failure ??= verdict === "unrelated" ? reject("delegation-invalid") : verdict;
  }
  return failure ?? reject("signer-mismatch");
}

function checkDelegation(
  candidate: Json,
  device: string,
  identity: string,
  documents: DidDocuments,
  ctx: ReceiverContext,
): "bound" | "unrelated" | Reject {
  const env = asEnvelope(candidate);
  const bytes = env && b64urlDecode(env.payload);
  const claimed = bytes ? decodePayload(bytes) : null;
  if (!claimed || isReject(claimed)) return "unrelated";
  if (claimed["device"] !== device || claimed["subject"] !== identity) return "unrelated";

  const opened = open(candidate, documents);
  if (isReject(opened)) return reject("delegation-invalid");
  if (claimed["type"] !== "DeviceDelegation" || didOf(opened.kid) !== identity) return reject("delegation-invalid");
  const { capabilities, issued_at, expires_at } = claimed;
  if (!Array.isArray(capabilities) || typeof issued_at !== "number" || typeof expires_at !== "number") {
    return reject("delegation-invalid");
  }
  if (!capabilities.includes("dsip.signaling")) return reject("delegation-capability");
  if (!(issued_at <= ctx.now && ctx.now < expires_at)) return reject("delegation-expired");
  const revoked = revocationsFor(identity, ctx).some((r) => r.device === device && issued_at <= r.revoked_at);
  return revoked ? reject("delegation-revoked") : "bound";
}

/**
 * Verify a received envelope: stages 1–14, first failure wins.
 *
 * `frame`, when given, is the exact text frame the envelope arrived in (§13.2 size cap).
 */
export function verifyEnvelope(envelope: Json, ctx: ReceiverContext, schemas: SchemaSet, frame?: string): Verdict {
  if (frame !== undefined && Buffer.byteLength(frame, "utf8") > MAX_FRAME_BYTES) {
    return reject("frame-too-large", "transport.envelope-too-large");
  }
  const opened = open(envelope, ctx.did_documents ?? {});
  if (isReject(opened)) return opened;
  const payload = decodePayload(opened.bytes);
  if (isReject(payload)) return payload;

  const { dsip, type, id, from, issued_at, expires_at } = payload;
  const shaped =
    !!dsip && typeof dsip === "object" && !Array.isArray(dsip) &&
    typeof type === "string" && typeof id === "string" && isUlid(id) && typeof from === "string" &&
    Number.isInteger(issued_at) && Number.isInteger(expires_at);
  if (!shaped) return reject("payload-shape");
  const issued = issued_at as number;
  const expires = expires_at as number;

  // Stage 7: bind kid → from, and on hello from → on_behalf_of (§7.4, §13.2).
  const signer = didOf(opened.kid);
  let identity = from as string;
  let unbound = signer === identity ? null : bind(signer, identity, opened.header, ctx);
  if (!unbound && type === "hello" && typeof payload["on_behalf_of"] === "string") {
    if (payload["on_behalf_of"] !== identity) unbound = bind(identity, payload["on_behalf_of"], opened.header, ctx);
    identity = payload["on_behalf_of"];
  }
  if (unbound) return type === "hello" ? reject(unbound.code, "transport.hello-rejected") : unbound;

  if (expires <= issued) return reject("expiry-order");
  if (type === "introduction") {
    // §12.9 (v0.8): a held introduction has no age bound; its validity is capped instead.
    if (expires - issued > MAX_INTRODUCTION_VALIDITY_S) return reject("introduction-validity");
    if (issued > ctx.now + REPLAY_WINDOW_S) return reject("replay-window");
  } else if (issued < ctx.now - REPLAY_WINDOW_S || issued > ctx.now + REPLAY_WINDOW_S) {
    return reject("replay-window");
  }
  if (expires < ctx.now) return reject("expired", type === "invite" ? "session.expired" : undefined);
  if ((ctx.seen_ids ?? []).includes(id as string)) return reject("duplicate-id");
  const ulidAt = ulidSeconds(id as string);
  if (ulidAt !== null && Math.abs(ulidAt - issued) > ULID_TOLERANCE_S) return reject("ulid-issued-at-mismatch");
  if (ctx.hello_verified === false && type !== "hello") return reject("hello-required", "transport.hello-required");

  const encoded = typeof envelope === "string" ? envelope : JSON.stringify(envelope);
  const verdict = checkPayload(
    payload,
    { ...ctx, signer_kid: opened.kid, encoded_size: ctx.encoded_size ?? Buffer.byteLength(encoded, "utf8") },
    schemas,
  );
  if (verdict.verdict === "reject") return verdict;
  return { ...verdict, type: type as string, signer, identity };
}
